pkg/wasm/cache.go

```diff
@@ -0,0 +1,194 @@
+		wasmModuleExpiry: moduleExpiry,
+		stopChan:         make(chan struct{}),
+	}
+	go func() {
+		cache.purge()
+	}()
+	return cache
+}
+
+// Get returns path the local Wasm module file.
+func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration) (string, error) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	url, err := url.Parse(downloadURL)
+	if err != nil {
+		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
+	}
+	// Construct Wasm cache key with downloading URL and provided checksum of the module.
+	key := cacheKey{
+		downloadURL: downloadURL,
+		checksum:    checksum,
+	}
+
+	switch url.Scheme {
+	case "http", "https":
+		// First check if the cache entry is already downloaded.
+		if ce, ok := c.modules[key]; ok {
+			// Update last touched time.
+			ce.last = time.Now()
+			return ce.modulePath, nil
+		}
+
+		// If the module is not available locally, download the Wasm module with http fetcher.
+		b, err := c.httpFetcher.Fetch(downloadURL, timeout)
+		if err != nil {
+			return "", err
+		}
+
+		// Get sha256 checksum and check if it is the same as provided one.
+		dChecksum := fmt.Sprintf("%x", sha256.Sum256(b))
+		if checksum != "" && dChecksum != checksum {
+			return "", fmt.Errorf("module downloaded from %v has checksum %v, which does not match: %v", downloadURL, dChecksum, checksum)
+		}
+
+		// TODO(bianpengyuan): Add sanity check on downloaded file to make sure it is a valid Wasm module.
+
+		key.checksum = dChecksum
+		// Check if the module has already been stored as local file. If so, avoid writing the file again.
+		if ce, ok := c.modules[key]; ok {
+			// Update last touched time.
+			ce.last = time.Now()
+			return ce.modulePath, nil
+		}
+
+		// Materialize the Wasm module into a local file. Use checksum as name of the module.
+		f := filepath.Join(c.dir, fmt.Sprintf("%s.wasm", dChecksum))
+		err = ioutil.WriteFile(f, b, 0644)
+		if err != nil {
+			return "", err
+		}
+
+		// Add the downloaded module into the cache map.
+		ce := cacheEntry{
+			modulePath: f,
+			last:       time.Now(),
+		}
+		c.modules[key] = ce
+
+		return f, nil
+	default:
+		return "", fmt.Errorf("unsupported Wasm module downloading URL scheme: %v", url.Scheme)
+	}
+}
+
+// Purge periodically clean up the stale Wasm modules local file and the cache map.
+func (c *LocalFileCache) purge() {
+	ticker := time.NewTicker(c.purgeInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			c.mux.Lock()
+			for k, m := range c.modules {
+				if m.expired(c.wasmModuleExpiry) {
+					// The module has not be touched for Expiry, delete if from the map as well as the local dir.
+					if err := os.Remove(m.modulePath); err != nil {
+						wasmLog.Errorf("failed to purge Wasm module %v: %v", m.modulePath, err)
+					} else {
+						delete(c.modules, k)
+						wasmLog.Debugf("successfully removed stale Wasm module %v", m.modulePath)
+					}
+				}
+			}
+			c.mux.Unlock()
+		case <-c.stopChan:
```

Comment:
nit: don't see where it's signaled.

Reply:
This is signaled in the test, so that multiple runs won't interfere each other.

I am not sure whether we'd want to signal it in real logic, we can just let it run for the lifetime of the proxy.